### PR TITLE
feat: Make parquet_file.partition_id optional in the catalog

### PR DESCRIPTION
**Pull request by carols10cents**
Connects to https://github.com/influxdata/idpe/issues/17476.

This will acquire a short lock on the table in postgres, per: <https://stackoverflow.com/questions/52760971/will-making-column-nullable-lock-the-table-for-reads>

This allows us to persist data for new partitions and associate the Parquet file catalog records with the partition records using only the partition hash ID, rather than both that are used now.

This will make https://github.com/influxdata/influxdb_iox/pull/8295 nicer.

### CODE REVIEW

**Comment by carols10cents** [comment]
ohhh noooo i got `TEST_INTEGRATION=1 cargo test -p iox_catalog` working and thought i was done oh nooooo

**Comment by carols10cents** [comment]
WOOOOO ok i think this is ready for review @domodwyer; in the meantime i'm going to rebase https://github.com/influxdata/influxdb_iox/pull/8295 on top of this and simplify it, hopefully.

**Comment by carols10cents** [comment]
@wiedld there are a few changes to compactor/scheduler tests in here. I'm happy to manage any merge conflicts. I'd also love to know if you think the changes make sense!

File: querier/src/parquet/query_access.rs [hunk 1]
```diff
@@ -15,11 +15,11 @@ impl QueryChunk for QuerierParquetChunk {
     }
 
     fn partition_id(&self) -> PartitionId {
-        self.meta().partition_id()
+        unimplemented!()
```
Comment (carols10cents) [inline querier/src/parquet/query_access.rs:18]: This gets resolved in https://github.com/influxdata/influxdb_iox/pull/8295 i swear. And hey, no tests fail, so it must be fine, right?

File: data_types/src/lib.rs [hunk 2]
```diff
@@ -527,10 +527,9 @@ pub struct ParquetFile {
     pub namespace_id: NamespaceId,
     /// the table
     pub table_id: TableId,
-    /// the partition
-    pub partition_id: PartitionId,
-    /// the partition hash ID, if generated
-    pub partition_hash_id: Option<PartitionHashId>,
```
Comment (domodwyer) [inline data_types/src/lib.rs:533]: This is looking good!

File: iox_catalog/migrations/20230726175943_make_parquet_file_partition_id_optional.sql [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+AS $$
+BEGIN
+    UPDATE partition
+    SET new_file_at = NEW.created_at
+    WHERE (NEW.partition_id IS NULL OR id = NEW.partition_id)
+      AND (NEW.partition_hash_id IS NULL OR hash_id = NEW.partition_hash_id);
+
+    RETURN NEW;
+END;
+$$;
```
Comment (domodwyer) [inline iox_catalog/migrations/20230726175943_make_parquet_file_partition_id_optional.sql:19]: I was very surprised we had this - why doesn't the application set this field?!

**Review by domodwyer — APPROVED** [review]

**Comment by domodwyer** [comment]
I believe the lock for the DDL change will be fast enough for us to not worry.

**Pull request merged**